pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+		if container.GetState() != runtimeapi.ContainerState_CONTAINER_RUNNING {
+			continue
+		}
+		id := container.GetId()
+		// Note that we should not block log rotate for an error of a single container.
+		status, err := c.runtimeService.ContainerStatus(id)
+		if err != nil {
+			glog.Errorf("Failed to get container status for %q: %v", id, err)
+			continue
+		}
+		path := status.GetLogPath()
+		info, err := os.Stat(path)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				glog.Errorf("Failed to stat container log %q: %v", path, err)
+				continue
+			}
+			// In rotateLatestLog, there are several cases that we may
+			// lose original container log after ReopenContainerLog fails.
+			// We try to to recover it by reopening container log.
+			if err := c.runtimeService.ReopenContainerLog(id); err != nil {
+				glog.Errorf("Container %q log %q doesn't exist, reopen container log failed: %v", id, path, err)
+				continue
+			}
+			// The container log should be recovered.
+			info, err = os.Stat(path)
+			if err != nil {
+				glog.Errorf("Failed to stat container log %q after reopen: %v", path, err)
+				continue
+			}
+		}
+		if info.Size() < c.policy.MaxSize {
+			continue
+		}
+		// Perform log rotation.
+		if err := c.rotateLog(id, path); err != nil {
+			glog.Errorf("Failed to rotate log %q for container %q: %v", path, id, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func (c *containerLogManager) rotateLog(id, log string) error {
+	// pattern is used to match all rotated files.
+	pattern := fmt.Sprintf("%s.*", log)
+	logs, err := filepath.Glob(pattern)
+	if err != nil {
+		return fmt.Errorf("failed to list all log files with pattern %q: %v", pattern, err)
+	}
+
+	logs, err = c.cleanupUnusedLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to cleanup logs: %v", err)
+	}
+
+	logs, err = c.removeExcessLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to remove excess logs: %v", err)
+	}
+
+	// Compress uncompressed log files.
+	for _, l := range logs {
+		if strings.HasSuffix(l, compressSuffix) {
+			continue
+		}
+		if err := c.compressLog(l); err != nil {
+			return fmt.Errorf("failed to compress log %q: %v", l, err)
+		}
+	}
+
+	if err := c.rotateLatestLog(id, log); err != nil {
+		return fmt.Errorf("failed to rotate log %q: %v", log, err)
+	}
+
+	return nil
+}
+
+// cleanupUnusedLogs cleans up temporary or unused log files generated by previous log rotation
+// failure.
+// logs passed in must be in increasing order.
+func (c *containerLogManager) cleanupUnusedLogs(logs []string) ([]string, error) {
+	inuse, unused := filterUnusedLogs(logs)
+	for _, l := range unused {
+		if err := os.Remove(l); err != nil {
+			return nil, fmt.Errorf("failed to remove unused log %q: %v", l, err)
+		}
+	}
+	return inuse, nil
+}
+
+// filterUnusedLogs splits logs into 2 groups, the 1st group is in used logs,
+// the second group is unused logs.
+func filterUnusedLogs(logs []string) ([]string, []string) {
```

Comment:
define inuse and unused in the function declaration:
`func filterUnusedLogs(logs []string) (inuse []string, unused []string)`

Reply:
Done.